epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.kt

```diff
@@ -0,0 +1,506 @@
+        val recyclerView = attachedRecyclerView
+        if (recyclerView != null) {
+            val itemAnimator = recyclerView.itemAnimator
+            if (checkItemAnimator && itemAnimator != null) {
+                // `itemAnimatorFinishedListener.onAnimationsFinished` will process visibility check
+                // - If the animations are running `onAnimationsFinished` will be invoked on animations end.
+                // - If the animations are not running `onAnimationsFinished` will be invoked right away.
+                if (itemAnimator.isRunning(itemAnimatorFinishedListener)) {
+                    // If running process visibility now as `onAnimationsFinished` was not yet called
+                    processChangeEventWithDetachedView(null, debug)
+                }
+            } else {
+                processChangeEventWithDetachedView(null, debug)
+            }
+        }
+    }
+
+    private fun processChangeEventWithDetachedView(detachedView: View?, debug: String) {
+        val recyclerView = attachedRecyclerView
+        if (recyclerView != null) {
+
+            // On every every events lookup for a new adapter
+            processNewAdapterIfNecessary()
+
+            // Process the detached child if any
+            if (detachedView != null) {
+                processChild(detachedView, true, debug)
+            }
+
+            // Process all attached children
+            for (i in 0 until recyclerView.childCount) {
+                val child = recyclerView.getChildAt(i)
+                if (child != null && child !== detachedView) {
+                    // Is some case the detached child is still in the recycler view. Don't process it as it
+                    // was already processed.
+                    processChild(child, false, debug)
+                }
+            }
+        }
+    }
+
+    /**
+     * If there is a new adapter on the attached RecyclerView it will resister the data observer and
+     * clear the current visibility states
+     */
+    private fun processNewAdapterIfNecessary() {
+        attachedRecyclerView?.adapter?.let { adapter ->
+            if (lastAdapterSeen != adapter) {
+                // Unregister the old adapter
+                lastAdapterSeen?.unregisterAdapterDataObserver(observer)
+                // Register the new adapter
+                adapter.registerAdapterDataObserver(observer)
+                lastAdapterSeen = adapter
+            }
+        }
+    }
+
+    /**
+     * Don't call this method directly, it is called from
+     * [EpoxyVisibilityTracker.processVisibilityEvents]
+     *
+     * @param child               the view to process for visibility event
+     * @param detachEvent         true if the child was just detached
+     * @param eventOriginForDebug a debug strings used for logs
+     */
+    private fun processChild(child: View, detachEvent: Boolean, eventOriginForDebug: String) {
```

Comment:
@eboudrant wanted to put this in a separate review so we could have threaded comments. My potentially naive solution was to replace this method with the following. Basically it checks if the child is a model group and processes each one, taking care to also add nested trackers for recycler views. Curious what your thoughts are here or if you have a more robust solution.

```
    private fun processChild(child: View, detachEvent: Boolean, eventOriginForDebug: String) {
        val recyclerView = attachedRecyclerView ?: return
        // Preemptive check for child's parent validity to prevent `IllegalArgumentException` in
        // `getChildViewHolder`.
        val isParentValid = child.parent == null || child.parent === recyclerView
        val viewHolder = if (isParentValid) recyclerView.getChildViewHolder(child) else null
        if (viewHolder is EpoxyViewHolder) {
            val epoxyHolder = viewHolder.holder
            if (epoxyHolder is ModelGroupHolder) {
                processModelGroupChildren(epoxyHolder, detachEvent, eventOriginForDebug)
            } else {
                processChild(child, detachEvent, eventOriginForDebug, viewHolder)
            }
        }
    }

    /**
     * Loop through the children of the model group and process visibility events on each one in
     * relation to the model group's layout. This will attach or detach trackers to any nested
     * [RecyclerView]s.
     *
     * @param epoxyHolder         the [ModelGroupHolder] with children to process
     * @param detachEvent         true if the child was just detached
     * @param eventOriginForDebug a debug strings used for logs
     */
    private fun processModelGroupChildren(
        epoxyHolder: ModelGroupHolder,
        detachEvent: Boolean,
        eventOriginForDebug: String
    ) {
        // Iterate through models in the group and process each of them instead of the group
        for (groupChildHolder in epoxyHolder.viewHolders) {
            // Since the group is likely using a ViewGroup other than a RecyclerView we need to
            // handle the potential of a nested RecyclerView.
            if (groupChildHolder.itemView is RecyclerView) {
                if (detachEvent) {
                    processChildRecyclerViewDetached(groupChildHolder.itemView)
                } else {
                    processChildRecyclerViewAttached(groupChildHolder.itemView)
                }
            }
            processChild(
                groupChildHolder.itemView,
                detachEvent,
                eventOriginForDebug,
                groupChildHolder
            )
        }
    }

    /**
     * Process visibility events for a view and propagate to a nested tracker if the view is a 
     * [RecyclerView].
     */
    private fun processChild(
        child: View,
        detachEvent: Boolean,
        eventOriginForDebug: String,
        viewHolder: EpoxyViewHolder
    ) {
        val recyclerView = attachedRecyclerView ?: return
        val changed = processVisibilityEvents(
            recyclerView,
            viewHolder,
            detachEvent,
            eventOriginForDebug
        )
        if (changed && child is RecyclerView) {
            val tracker = nestedTrackers[child]
            tracker?.processChangeEvent("parent")
        }
    }
```

Reply:
Agree, let's convert first to Kotlin and create new review for everything related to group / visibility. but looking at your proposal, I like it.

Reply:
@pmecho I tried your code and it seems to work. One change I needed is to remove the `else` in `processChild` : 
```
            processChild(child, detachEvent, eventOriginForDebug, viewHolder)
            if (epoxyHolder is ModelGroupHolder) {
                processModelGroupChildren(epoxyHolder, detachEvent, eventOriginForDebug)
            }
```

How about you open a new pull request with your change and we iterate from there? I'd like to do more testing.

Reply:
@eboudrant Oh nice catch. Probably makes sense to still call the visibility methods on the model group itself. I made that change and pushed a draft PR: https://github.com/airbnb/epoxy/pull/1091